src/box/memtx_tx.c

```diff
@@ -3659,6 +3663,55 @@ memtx_tx_snapshot_clarify_slow(struct memtx_tx_snapshot_cleaner *cleaner,
 	return tuple;
 }
 
+size_t
+memtx_tx_snapshot_invisible_count_matching_until_slow(
+	struct memtx_tx_snapshot_cleaner *cleaner, struct index_def *index_def,
+	enum iterator_type type, const char *key, uint32_t part_count,
+	struct tuple *until, hint_t until_hint)
+{
+	assert(cleaner->ht != NULL);
+
+	struct key_def *key_def = index_def->key_def;
+	struct key_def *cmp_def = index_def->cmp_def;
+
+	/*
+	 * The border is only valid if it's located at or after the first
+	 * tuple in the index according to the iterator direction and key.
+	 */
+	assert(until == NULL ||
+	       memtx_tx_tuple_matches(key_def, until, type == ITER_EQ ?
+				      ITER_GE : type == ITER_REQ ? ITER_LE :
+				      type, key, part_count));
+
+	size_t res = 0;
+	struct mh_snapshot_cleaner_t *ht = cleaner->ht;
+	mh_int_t i;
+	mh_foreach(ht, i) {
+		struct memtx_tx_snapshot_cleaner_entry *entry =
+			mh_snapshot_cleaner_node(ht, i);
+
+		/* Not a tuple with a top story. */
+		if (!entry->in_pk)
+			continue;
+
+		/* Top story, but the tuple is excluded from the index. */
+		if (tuple_key_is_excluded(entry->from, key_def, MULTIKEY_NONE))
+			continue;
+
+		/* In-index tuple, but does not match with key and border. */
+		if (!memtx_tx_tuple_matches_until(key_def, entry->from, type,
+						  key, part_count, cmp_def,
+						  until, until_hint))
+			continue;
+
+		/* Count the invisible tuple. */
+		if (entry->to == NULL ||
+		    memtx_tx_snapshot_clarify_slow(cleaner, entry->to) == NULL)
```

Comment:
I fail to understand why you both iterate over *all* hash table entries and do lookups along the way.

Reply:
That's done cause the tuple clarification routine clarifies tuples in a loop. IDK what was the original point in this loop, but after introducing alter statements it started to make sense cause the altered tuple might not point directly to the cleared tuple, so we may need one more iteration. We don't care about alter statements here, so I removed the call. Also only handled the alter stamtements in no-MVCC mode, since it's not required in this case, see the "memtx: only store in-index tuples in the cleaner" commit.

Thanks.